Firestore/core/test/unit/bundle/bundle_serializer_test.cc

```diff
@@ -26,10 +38,165 @@ namespace bundle {
+    document.mutable_update_time()->set_nanos(now.nanoseconds());
+    document.mutable_update_time()->set_seconds(now.seconds());
+    document.mutable_fields()->insert({"foo", value});
+
+    return document;
+  }
+
+  // 1. Take the value, put it in a libprotobuf Document message and print it
+  // into a JSON string.
+  // 2. Use BundleSerializer to parse the string, then encode the parsed
+  // document into nanopb bytes.
+  // 3. Parse the nanopb bytes to libprotobuf Document message, then compare
+  // with the original.
+  void VerifyFieldValueRoundtrip(ProtoValue value) {
+    ProtoDocument document = TestDocument(value);
+
+    std::string json_string;
+    MessageToJsonString(document, &json_string);
+
+    auto actual = VerifyJsonStringDecodes(json_string);
+
+    VerifyDecodedDocumentEncodesToOriginal(actual.document(), document);
+  }
+
+  void VerifyDecodedDocumentEncodesToOriginal(const model::Document& decoded,
+                                              const ProtoDocument& original) {
+    ByteString bytes =
+        nanopb::MakeByteString(local_serializer.EncodeMaybeDocument(decoded));
+    ProtoMaybeDocument maybe_document;
+    *maybe_document.mutable_document() = original;
+    EXPECT_TRUE(msg_diff_.Compare(maybe_document,
+                                  ProtobufParse<ProtoMaybeDocument>(bytes)))
+        << message_differences;
+    message_differences.clear();
+  }
+
+  void VerifyFieldValueDecodeFails(ProtoValue value) {
+    ProtoDocument document = TestDocument(value);
+
+    std::string json_string;
+    MessageToJsonString(document, &json_string);
+    VerifyJsonStringDecodeFails(std::move(json_string));
+  }
+
+  BundleDocument VerifyJsonStringDecodes(std::string json_string) {
+    JsonReader reader;
+    BundleDocument actual =
+        bundle_serializer.DecodeDocument(reader, json_string);
+    EXPECT_TRUE(reader.ok());
+    return actual;
+  }
+
+  void VerifyJsonStringDecodeFails(std::string json_string) {
+    JsonReader reader;
+    BundleDocument actual =
+        bundle_serializer.DecodeDocument(reader, json_string);
+    EXPECT_FALSE(reader.ok());
+  }
+
+  // 1. Take a `Query` object, put it in a `NamedQuery` and encode it to byte
+  // array via nanopb.
+  // 2. Parse the byte array to libprotobuf named query
+  // 3. Get Json presentation of the protobuf named query
+  // 4. Parse the json back to `NamedQuery` object, then compare.
+  void VerifyNamedQueryRoundtrip(const Query& original) const {
+    // `First` and `None` will be encoded as BundledQuery.limit_type = `First`,
+    // as not all SDKs have a `None`.
+    auto limit_type = original.limit_type() == core::LimitType::Last
+                          ? core::LimitType::Last
+                          : core::LimitType::First;
+    BundledQuery bundled_query(original.ToTarget(), limit_type);
+    NamedQuery named_query("query-1", bundled_query,
+                           model::SnapshotVersion(Timestamp::Now()));
```

Comment:
`testutil::Version` is more concise than creating an actual snapshot version. It also results in a specific version, which will make diffs between test runs less random.

Reply:
Done.